testing.rst

```diff
@@ -51,79 +50,111 @@ Symfony application.
     Use the ``--coverage-*`` command options to generate code coverage reports.
     Read the PHPUnit manual to learn more about `code coverage analysis`_.
 
-.. index::
-   single: Tests; Unit tests
+Types of Tests
+--------------
+
+`Unit Tests`_
+    These tests ensure that *individual* units of source code (e.g. a single
+    class) behave as intended.
+
+`Integration Tests`_
+    These tests test a combination of classes and commonly interact with
+    Symfony's service container. These tests do not yet cover the full
+    working application, those are called *Functional tests*.
+
+`Functional Tests`_
+    Functional tests test the behavior of a complete application. They
+    make HTTP requests and test that the response is as expected.
+
+`End to End Tests (E2E)`_
+    At last, end to end tests test the application as a real user. They use
+    a real browser and real integrations with external services.
 
 Unit Tests
 ----------
 
-A `unit test`_ ensures that individual units of source code (e.g. a single class
-or some specific method in some class) meet their design and behave as intended.
-If you want to test an entire feature of your application (e.g. registering as a
-user or generating an invoice), see the section about :ref:`Functional Tests <functional-tests>`.
+A `unit test`_ ensures that individual units of source code (e.g. a single
+class or some specific method in some class) meet their design and behave
+as intended. Writing Symfony unit tests is no different from writing
+standard PHPUnit unit tests. You can learn about it in the PHPUnit
+documentation: `Writing Tests for PHPUnit`_.
 
-Writing Symfony unit tests is no different from writing standard PHPUnit
-unit tests. Suppose, for example, that you have an class called ``Calculator``
-in the ``src/Util/`` directory of the app::
+By convention, the ``tests/`` directory should replicate the directory
+of your application for unit tests. So, if you're testing a class in the
+``src/Util/`` directory, put the test in the ``tests/Util/`` directory.
```

Comment:
With the 4 test types officially defined, should we suggest an alternate `tests/` directory structure?
- `tests/Unit/`
- `tests/Integration/`
- `tests/Functional/`
- `tests/E2E/`

Reply:
I know that we never discussed this. But I also use the structure defined by Kevin. 

The benefit is that you can easily run your slow E2E tests less often.